akka-remote/src/main/scala/akka/remote/artery/FlightRecorder.scala

```diff
@@ -0,0 +1,336 @@
+    }
+
+    writeRecord()
+  }
+
+  def markSnapshot(logId: Int): Unit = buffers(logId).putLongVolatile(LogStateOffset, Snapshot)
+  def markLive(logId: Int): Unit = buffers(logId).putLongVolatile(LogStateOffset, Live)
+
+  def close(): Unit = buffers.foreach(_.close())
+}
+
+/**
+ * INTERNAL API
+ */
+private[remote] object FlightRecorder {
+  val Alignment = 64 * 1024 // Windows is picky about mapped section alignments
+
+  val MagicString = 0x31524641 // "AFR1", little-endian
+  val GlobalSectionSize = BitUtil.align(24, Alignment)
+  val StartTimeStampOffset = 4
+
+  val LogHeaderSize = 16
+  val SnapshotCount = 4
+  val SnapshotMask = SnapshotCount - 1
+
+  // TODO: Dummy values right now, format is under construction
+  val AlertRecordSize = 128
+  val LoFreqRecordSize = 128
+  val HiFreqBatchSize = 62
+  val HiFreqRecordSize = 16 * (HiFreqBatchSize + 2) // (batched events + header)
+
+  val AlertWindow = 256
+  val LoFreqWindow = 256
+  val HiFreqWindow = 256 // This is counted in batches !
+
+  val AlertLogSize = BitUtil.align(LogHeaderSize + (AlertWindow * AlertRecordSize), Alignment)
+  val LoFreqLogSize = BitUtil.align(LogHeaderSize + (LoFreqWindow * LoFreqRecordSize), Alignment)
+  val HiFreqLogSize = BitUtil.align(LogHeaderSize + (HiFreqWindow * HiFreqRecordSize), Alignment)
+
+  val AlertSectionSize = AlertLogSize * SnapshotCount
+  val LoFreqSectionSize = LoFreqLogSize * SnapshotCount
+  val HiFreqSectionSize = HiFreqLogSize * SnapshotCount
+
+  val AlertSectionOffset = GlobalSectionSize
+  val LoFreqSectionOffset = GlobalSectionSize + AlertSectionSize
+  val HiFreqSectionOffset = GlobalSectionSize + AlertSectionSize + LoFreqSectionSize
+
+  val TotalSize = GlobalSectionSize + AlertSectionSize + LoFreqSectionSize + HiFreqSectionSize
+
+  val HiFreqEntryCountFieldOffset = 16
+}
+
+/**
+ * INTERNAL API
+ */
+private[akka] class FlightRecorder(val fileChannel: FileChannel) extends AtomicBoolean {
+  import FlightRecorder._
+
+  private[this] val globalSection = new MappedResizeableBuffer(fileChannel, 0, GlobalSectionSize)
+
+  // FIXME: check if power of two
+  private[this] val SnapshotMask = SnapshotCount - 1
+  private[this] val alertLogs =
+    new RollingEventLogSection(
+      fileChannel = fileChannel,
+      offset = AlertSectionOffset,
+      entryCount = AlertWindow,
+      logBufferSize = AlertLogSize,
+      recordSize = AlertRecordSize)
+  private[this] val loFreqLogs =
+    new RollingEventLogSection(
+      fileChannel = fileChannel,
+      offset = LoFreqSectionOffset,
+      entryCount = LoFreqWindow,
+      logBufferSize = LoFreqLogSize,
+      recordSize = LoFreqRecordSize)
+  private[this] val hiFreqLogs =
+    new RollingEventLogSection(
+      fileChannel = fileChannel,
+      offset = HiFreqSectionOffset,
+      entryCount = HiFreqWindow,
+      logBufferSize = HiFreqLogSize,
+      recordSize = HiFreqRecordSize)
+  // No need for volatile, guarded by atomic CAS and set
+  @volatile private var currentLog = 0
```

Comment:
I think it is not needed *if* I change the `set(false)` at L252 to a `compareAndSet(true, false)`. The latter is a full barrier, while the `set()` theoretically needs a matching volatile read for things to become visibe. Not relevant on x86/x64 but hey. That will allow me to get rid of that volatile.